python/cuspatial/cuspatial/core/binpreds/binpreds.py

```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
+            # Not necessary for points.
+            df_result = (
+                result.groupby("idx").sum().sort_index()
+                == result.groupby("idx").count().sort_index()
+            )
+        point_result = cudf.Series(
+            df_result["pip"], index=cudf.RangeIndex(0, len(df_result))
+        )
+        point_result.name = None
+        return point_result
+
+
+class EqualsBinpred(BinaryPredicate):
+    def _offset_equals(self, lhs, rhs):
+        """Compute the pairwise length equality of two offset arrays"""
+        lhs_lengths = lhs[:-1] - lhs[1:]
+        rhs_lengths = rhs[:-1] - rhs[1:]
+        return lhs_lengths == rhs_lengths
+
+    def _sort_multipoints(self, lhs, rhs, initial):
+        """Sort xy according to bins defined by offset"""
+        sort_indices = cp.repeat(lhs.point_indices(), 2)
+        lhs_xy, rhs_xy = lhs.xy, rhs.xy
+        lhs_xy.index = sort_indices
+        rhs_xy.index = sort_indices
+        lhs_df = lhs_xy.reset_index(drop=False, name="xy")
+        lhs_df["axis"] = lhs_df.index % 2
+        rhs_df = rhs_xy.reset_index(drop=False, name="xy")
+        rhs_df["axis"] = rhs_df.index % 2
+        lhs_sorted = lhs_df.sort_values(
+            by=["index", "axis", "xy"]
+        ).reset_index(drop=True)
+        rhs_sorted = rhs_df.sort_values(
+            by=["index", "axis", "xy"]
+        ).reset_index(drop=True)
+        return (
+            PreprocessorOutput(lhs_sorted["xy"], lhs.point_indices()),
+            PreprocessorOutput(rhs_sorted["xy"], rhs.point_indices()),
+            initial,
+        )
+
+    def _swap_linestring_first_and_last(self, linestring):
+        """Swap the first and last values of a linestring"""
+        # Save temporary values since x cannot be used unmodified
+        x = linestring.x
+        y = linestring.y
+        point_range = cudf.Series(cp.arange(len(x)))
+        indices = point_range.groupby(linestring.point_indices())
+        # Create masks for the first and last values of each linestring
+        swap_x = x[indices.last()].reset_index(drop=True) < x[
+            indices.first()
+        ].reset_index(drop=True)
+        swap_y = y[indices.last()].reset_index(drop=True) < y[
+            indices.first()
+        ].reset_index(drop=True)
+        # Swap the first and last values of each linestring
+        (x.iloc[indices.last()[swap_x]], x.iloc[indices.first()[swap_x]],) = (
+            x.iloc[indices.first()[swap_x]],
+            x.iloc[indices.last()[swap_x]],
+        )
+        (y.iloc[indices.last()[swap_y]], y.iloc[indices.first()[swap_y]],) = (
+            y.iloc[indices.first()[swap_y]],
+            y.iloc[indices.last()[swap_y]],
+        )
+        return cudf.Series(cp.column_stack((x.values, y.values)))
+
+    def _sort_linestrings(self, lhs, rhs, initial):
+        """Swap first and last values of each linestring to ensure that
+        the first point is the lowest value. This is necessary to ensure
```

Comment:
I can't seem to reply to all of the comments so I'll address them here.

1. It is possible with a pile of conditional logic to compute the group sizes via their offsets buffers, but using `object.point_indices()` and `groupby` allows me to avoid implementing those code paths. It is a "time of development vs. time of execution" problem, and the overall cost to a python user is not significant I think.

2. Not sure what you mean.

3. Summing a bool column with 255 values in `cudf` returns an `int64`.

4. It is lowest in the context of the endpoints, the first point needs to be lexicographically before the last point in order for comparison of all the points via sorting to work. I updated the docs.

Reply:
Can you move replies to: https://github.com/rapidsai/cuspatial/pull/926#discussion_r1134753854 ?

Reply:
And to your reply on this comment:

> It is lowest in the context of the endpoints, the first point needs to be lexicographically before the last point in order for comparison of all the points via sorting to work. I updated the docs.

Can you elaborate? How does the endpoints lexicographic order affects the equality of two linestrings?

Reply:
It doesn't! Thanks for calling out another mistake. I had built some test cases around strings of length 2 and ended up believing the data had to do with the start and end of the strings. What really happens is that a linestring is equal to the opposite linestring, which works with sort, but doesn't work with linestrings constructed of the same point in different orders.

The updated code now reverses each linestring within its "bucket" along the GeoArrow x/y buffer, then does a forward and a reverse comparison and `or`s them.